pkg/resource/deploy/source_eval.go

```diff
@@ -1115,6 +1131,13 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 
 	// If this is a remote component, fetch its provider and issue the construct call. Otherwise, register the resource.
 	var result *RegisterResult
+	if !custom {
+		defer func() {
```

Comment:
Why is this defered? Can't we always just store the provider list as it comes in?

Reply:
We know the value when we declare the deferred function (L1135), but we don't know the key (URN) to store it at. Note that the defer only finishes the store if we successfully stored an URN.

Reply:
Should we just move this code down to L1197 when we have the result, so we don't have to have the defer here?

If there's a reason it needs to live here with the defer, it needs a comment explaining why.